### ISSUE TITLE
attach shell (right-click) fails; it doesn't use docker host (aka docker.host setting)

### ISSUE THREAD

**Issue by diablodale** [microsoft/vscode-docker#1524]
Right-click container **Attach Shell** fails to attach shell because it doesn't use the `docker.host` setting on the command.

### Local Client

* Windows 10 x64
* VSCode
  Version: 1.41.1 (system setup)
  Commit: 26076a4de974ead31f97692a0d32f90d735645c0
  Date: 2019-12-18T14:58:56.166Z
  Electron: 6.1.5
  Chrome: 76.0.3809.146
  Node.js: 12.4.0
  V8: 7.6.303.31-electron.0
  OS: Windows_NT x64 10.0.18363
* VSCode Extension -> Docker
* VSCode Extension -> Remote Development
* WSL Ubuntu 18.04.3 LTS
* SSH tunnel to remote server using https://code.visualstudio.com/docs/remote/troubleshooting#_using-an-ssh-tunnel-to-connect-to-a-remote-docker-host

### Remote Server

* Ubuntu 18.04.3 LTS x86_64
* Docker Server 19.03.5
* some running containers

## Repro

1. Setup client and server as above.
2. Start the client ssh tunnel to the remote server as in the instructions above
3. Change the VSCode setting `"docker.host"` to match that ssh tunnel
4. Click on the docker whale icon in the left nav
5. Right-click on any container
6. Choose 'Attach Shell'

### Result

Attach fails. You get an error

```
docker exec -it abc123 /bin/sh -c "[ -e /bin/bash ] && /bin/bash || /bin/sh"

Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

### Expected

A shell prompt from inside the container

## Notes

The base command is incomplete. It needs to use the `-H` parameter to specify the host. It can not assume the host is local.

Errant `docker exec -it abc123`

Correct `docker -H proto://dockerhost exec -it abc123`

`proto://dockerhost` value should be retrieved from the `"docker.host"` VSCode setting.

**Comment by bwateratmsft** [comment]
For what it's worth, it is no longer necessary to use an SSH tunnel, you can directly use an SSH Docker host, starting with extension version 0.9.0. You will need to set up agent authentication; more info here: https://github.com/microsoft/vscode-docker/wiki/SSH

Nevertheless I'll investigate why this isn't working. We are supposed to be setting anything in `docker.host` into the `DOCKER_HOST` environment variable before executing any command lines.

**Comment by bwateratmsft** [comment]
@diablodale I wasn't able to reproduce your issue in a quick test. I was able to successfully use `Attach Shell` through an SSH tunnel. It's possible VSCode needed a reload. Can you try [using SSH directly](https://github.com/microsoft/vscode-docker/wiki/SSH) (don't forget to enable and configure the `ssh-agent` windows service), unless a tunnel is strictly necessary? If you have any trouble, let us know.

**Comment by diablodale** [comment]
Hi. I'm not clear what you are requesting of me.
1. You are deviating from the repo scenario. (e.g. running windows ssh-agent service). I have no comment on deviations.
2. I'm very sure everything is as in my repo. Docker is working well both locally and remotely as described in the setup above.
3. I already isolated the code fault in my notes.

I recommend you revisit this -- slowly. Please, follow the setup and repo steps. Don't deviate from them.

I reproduced the same errant behavior today with:
VSCode 1.42.0
Docker 0.10.0
Remote Development 0.19.0

Same error and same cause. The code continues to not clearly state the host. You must use the `-H` parameter.

```bash
docker exec -it 13c2b4b5628c373d577e49caa566a532eade5430229047c9d0257ade40cd79a0 /bin/sh -ccd79a0 /bin/sh -c "[ -e /bin/bash ] && /bin/bash || /bin/sh"
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

**Comment by bwateratmsft** [comment]
Hey @diablodale,

I tried both with and without the ssh-agent service, and it worked both ways. So, I did not deviate.

I read your notes thoroughly but using the `-H` flag is intentionally not the approach we are using. Our code takes VSCode settings for `docker.host` (and similar settings), and applies them to the `DOCKER_HOST` environment variable (and similar variables) for terminals created to run commands--see [here](https://github.com/microsoft/vscode-docker/blob/master/src/utils/addDockerSettingsToEnv.ts) and [here](https://github.com/microsoft/vscode-docker/blob/master/src/utils/TerminalProvider.ts). Because `DOCKER_HOST` is therefore set to the `tcp://...` or `ssh://...` values, `-H` isn't needed on the command line.

What I'm asking is for you to try directly using the `ssh://...` value as `DOCKER_HOST` (or `docker.host` setting in VSCode), rather than an SSH tunnel, as tunnels are no longer necessary. However, with that approach, it is necessary to configure an SSH agent as per the [wiki doc](https://github.com/microsoft/vscode-docker/wiki/SSH#directly-via-ssh), because unfortunately the `ssh2` node package used by `dockerode` does not work without one.

**Comment by diablodale** [comment]
Unfortunately, its not working as you intend. right-click attach as in my repo steps...plus an echo.

```bash
dale@dale-euro:/mnt/c/Users/Dale$ docker exec -it 13c2b4b5628c373d577e49caa566a532eade5430229047c9d0257ade40cd79a0 /bin/sh -c "[ -e /bin/bash ] && /bin/bash || /bin/sh"
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
dale@dale-euro:/mnt/c/Users/Dale$ echo $DOCKER_HOST 

dale@dale-euro:/mnt/c/Users/Dale$
```

There are probably dozens of ways a person can bridge communications between computers. Naturally, I have long-since fixed this for my specific setup. Separate from my ability to fix it for myself, is the global issue that can affect an unknown number of customers less savy. This issue is focused on using the SSH tunnel as I described. And with that scenario and repo steps, it continues to fail.

**Comment by bwateratmsft** [comment]
That blank value explains why the error is appearing. What shell are you using? That could be behaving different from the default cmd.exe on Windows.

**Comment by diablodale** [comment]
Bash. My shell in VSCode is always bash. Naturally, it could be foo, bar, CMD, csh, sh, zsh, or any number of other shells. Luckily `-H` parameter works for all of them. ;-)

Only designing and testing this feature for powershell is far too limiting.

**Comment by bwateratmsft** [comment]
We do design and test for other shells, including bash, sh, zsh, cmd, and PowerShell.

WSL is not at all the same thing as bash. You're running bash within the WSL environment, not the local environment, and it does not inherit environment variables from the host (and more importantly, from VSCode). Based on your setup, Docker is running on another machine, not within WSL, so I'm not sure why you'd want to add a layer by working through WSL.

We're using an official VSCode API for setting the [terminal environment variables](https://code.visualstudio.com/api/references/vscode-api#TerminalOptions); consequently any other extensions depending on that capability would also run into problems in this scenario. Surely not all of them could actually be solved with arguments. Asking all those extensions to work around that API, so that WSL can be used like it's just bash, is not the right approach. If you want you can ask the VSCode team to get those environment variables passed along into the bash session in WSL.

**Comment by diablodale** [comment]
@EricJizbaMSFT, I see you are the dev that committed the code to put docker settings into env vars. Would you please look at this issue? That code is failing to set the DOCKER_HOST env var when WSL is chosen as the integrated terminal shell.

@bwateratmsft , great to hear. Please do share with me how you have bash as your integrated terminal shell in a way that your code works? It must be one of those test cases you describe. I can then test that here to explore what the distinction is and we together can resolve from there. Of course, keeping in mind that calling BASH.EXE directly is deprecated -- Microsoft recommends calling WSL.EXE instead.

Additionally, I'm curious how you don't have WSL as one of your test cases? If it was, it would have already failed and your team know/escalate. WSL is an official supported integrated terminal shell. I can open any integrated terminal in VSCode...click the `1: xxxx` at the top-right of that terminal, and `select default shell`. WSL is an official listing there. I believe that all the VSCode teams should support all 4 of the officially supported terminal shells.

Workaround? Personally, I don't see it as that. Instead, a required part of defining the namespace and identifier for a container. Your intention is to launch into a docker container. docker.exe needs to know it is interactive, the namespace/location for the container (aka the docker engine host), the container ID, and the command to run. All of those are common docker.exe parameters. As the code is currently written you put some of that as params, some as env vars, and due to the split are running into a bug/lack of feature.

Lets see what @EricJizbaMSFT thinks of this and their input into the conversation.

**Comment by bwateratmsft** [comment]
If you want, you can use the `docker.attachShellCommand.linuxContainer` setting to hard-code a `-H proto://host` value into either workspace or user settings. In any case, expecting every extension using the `TerminalOptions` API to change their code in order to accommodate your abuse of WSL is not the right approach.

Assuming WSL is actually a supported shell, then `TerminalOptions.env` should work. I've filed https://github.com/microsoft/vscode/issues/90527 to have the VSCode team investigate.

**Comment by diablodale** [comment]
I appreciate you opening an issue with VSCode. However, there is still some investigation with your code that you didn't reply. As a reminder:

* share with me how you have bash as your integrated terminal shell in a way that your code works. You said bash was tested, then how did you team set that up and test it?
* You implicitly expressed that your team doesn't consider WSL a valid shell in VSCode, and therefore don't have it as one of your test cases. Isn't this a concern? @fiveisprime, you are the PM that wrote the blog "WSL 2 with Visual Studio Code". What do you think of this? 🤔

**Comment by bwateratmsft** [comment]
- We test bash on Windows with Git bash, and also on Mac and Linux with native bash. Git bash on Windows requires a small setting adjustment (#462) though there has not been enough demand to justify making a code change.
- WSL is either a valid shell, in which case the VSCode team needs to fix https://github.com/microsoft/vscode/issues/90527, or it is not a valid shell, in which case we have nothing to fix. It is not our decision to make whether or not it should be supported.

If indeed they decide to fix it, it should automatically fix this issue (and we will test to be sure). In the meantime, I've offered you multiple [solutions](https://github.com/microsoft/vscode-docker/issues/1524#issuecomment-584758360) and [workarounds](https://github.com/microsoft/vscode-docker/issues/1524#issuecomment-585212478) to what is a very corner case and you've been rude the entire time.

**Comment by EricJizbaMSFT** [comment]
I'm not entirely following along, but @fiveisprime and I do not use `terminal.integrated.shell.windows` for wsl - we use the [Remote WSL extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-wsl) instead. As its main page says:

> Remote - WSL runs commands and extensions directly in WSL so you don't have to worry about pathing issues, binary compatibility, or other cross-OS challenges.

I believe all of this environment variable frustration fits under the "challenges" used as a motivation for the Remote WSL extension.

@fiveisprime do we need to update the docs to make it more explicit that people should use the Remote WSL extension instead of the integrated terminal setting? For example [here](https://code.visualstudio.com/docs/editor/integrated-terminal#_can-i-use-the-integrated-terminal-with-the-windows-subsystem-for-linux)

**Comment by diablodale** [comment]
@bwateratmsft I'm not engaging in your defensive/triggering language towards me. I've escalated your behavior as a violation of Code of Conduct for review.
Now back to the technical discussion...

Thanks @EricJizbaMSFT for that link. That is another place showing that WSL is an officially supported and documented integrated terminal choice. This again highlights the lack of test cases in this extension's test pipeline for WSL as the customer's choice.

Of course, this extension could choose to not support WSL, in which case that 'shell' right-click action could be disabled/removed when the customer has chosen WSL, or this bug can just continue to live forever and not be fixed. This discussion can be available for people to find when they encounter problems where they can (at least) learn it is a known issue and there is no planned fix.

Yes to you both, above we already discussed are other ways to connect to remote hosts and ways to use WSL. @EricJizbaMSFT , the remote-WSL extension doesn't help in this scenario because this extension's code doesn't use the remote-WSL extension's abilities. Instead, it is my understanding from the above posts that this extension directly uses the integrated terminal shell the customer has selected (currently 4 integrated shells are natively supported by vscode, see my two posts above). @bwateratmsft doesn't write of any WSL detection or handling that could use/bounce into the remote-WSL extension.
